src/dmd/dtoh.d

```diff
@@ -0,0 +1,1256 @@
+    buf.writestring("#include <stdio.h>\n");
+    buf.writestring("#include <string.h>\n");
+    buf.writeByte('\n');
+    buf.writestring("#define _d_void void\n");
+    buf.writestring("#define _d_bool bool\n");
+    buf.writestring("#define _d_byte signed char\n");
+    buf.writestring("#define _d_ubyte unsigned char\n");
+    buf.writestring("#define _d_short short\n");
+    buf.writestring("#define _d_ushort unsigned short\n");
+    buf.writestring("#define _d_int int\n");
+    buf.writestring("#define _d_uint unsigned\n");
+    if (global.params.isLP64)
+    {
+        buf.writestring("#define _d_long long\n");
+        buf.writestring("#define _d_ulong unsigned long\n");
+    }
+    else
+    {
+        buf.writestring("#define _d_long long long\n");
+        buf.writestring("#define _d_ulong unsigned long long\n");
+    }
+    buf.writestring("#define _d_float float\n");
+    buf.writestring("#define _d_double double\n");
+    buf.writestring("#define _d_real long double\n");
+    buf.writestring("#define _d_char char\n");
+    buf.writestring("#define _d_wchar wchar_t\n");
+    buf.writestring("#define _d_dchar unsigned\n");
+    buf.writestring("\n");
+    buf.writestring("#define _d_null NULL\n");
+    buf.writestring("\n");
+    buf.writestring("struct AA;\n");
+    buf.writestring("\n");
+
+    OutBuffer check;
+    check.writestring(`
+#if OFFSETS
+
+template <class T>
+size_t getSlotNumber(int dummy, ...)
+{
+    T c;
+    va_list ap;
+    va_start(ap, dummy);
+    void *f = va_arg(ap, void*);
+    for (size_t i = 0; ; i++)
+    {
+        if ( (*(void***)&c)[i] == f)
+            return i;
+    }
+    va_end(ap);
+}
+
+void testOffsets()
+{
+    `);
+
+    OutBuffer done;
+    OutBuffer decl;
+    scope v = new ToCppBuffer!ASTCodegen(&check, buf, &done, &decl);
+    foreach (m; *ms)
+    {
+        buf.printf("// Parsing module %s\n", m.toPrettyChars());
+        m.accept(v);
+    }
+    buf.write(&done);
+    buf.write(&decl);
+
+    check.writestring(`
+}
+#endif
+    `);
+
+    debug buf.write(&check);
+}
+
+void main()
+{
+    import std.stdio;
+    import std.algorithm.sorting : sort;
+    import std.array : array;
+    import std.file : readText;
+    import std.path : baseName, buildPath, dirName;
+    import std.string : toStringz;
+
+    import dmd.id;
+    import dmd.parse;
+    import dmd.dsymbolsem;
+    import dmd.semantic2;
+    import dmd.semantic3;
+    import dmd.builtin : builtin_init;
+    import dmd.dmodule : Module;
+    import dmd.cond : VersionCondition;
+    import dmd.expression : Expression;
+    import dmd.objc : Objc;
+    import dmd.target : Target;
+
+    import core.memory;
+
+    GC.disable();
+
+    global._init();
+    global.params.isLinux = true;
+    global.params.is64bit = (size_t.sizeof == 8);
+    global.params.isLP64 = global.params.is64bit;
+
+    global.path = new Strings();
+    global.path.push(__FILE_FULL_PATH__.dirName.buildPath("../../../druntime/src/").toStringz());
```

Comment:
This is fragile if `druntime` is checkout out in a different directory (I think the LDC people do this).

Reply:
Yes our path is `$LDCROOT/runtime/druntime`.